saq/queue.py

```diff
@@ -393,12 +412,122 @@ async def enqueue(self, job_or_func, **kwargs):
+        ]
+        for gen in on_enqueue_generators:
+            await run_iterator_to_first_yield(gen)
+
+        enqueued = await self._enqueue_script(
             keys=[self._incomplete, job.id, self._queued, job.abort_id],
             args=[self.serialize(job), job.scheduled],
             client=self.redis,
-        ):
-            return None
+        )
+
+        for gen in on_enqueue_generators:
+            await run_iterator_to_completion(gen)
+
+        if enqueued:
+            logger.info("Enqueuing %s", job)
+            return job
+        return None
+
+    async def apply(self, job_or_func, timeout=None, **kwargs):
+        """
+        Enqueue a job and wait for its result.
+
+        If the job is successful, this returns its result.
+        If the job is unsuccessful, this raises a JobError.
+
+        Example::
+            try:
+                assert await queue.apply("add", a=1, b=2) == 3
+            except JobError:
+                print("job failed")
+
+        job_or_func: Same as Queue.enqueue
+        timeout: Seconds to wait for job to complete. If None or 0, wait forever.
+        kwargs: Same as Queue.enqueue
+        """
+        job = await self.enqueue(job_or_func, **kwargs)
+        return await job.wait_for_result(timeout)
+
+    async def map(self, job_or_func, iter_kwargs, timeout=None, **kwargs):
+        """
+        Enqueue multiple jobs and collect all of their results.
+
+        This raises a JobError if any of the jobs are unsuccessful.
+
+        This waits for job results in the order they are enqueued, so JobExceptions are
+        not necessarily raised immediately.
+
+        Example::
+            try:
+                assert await queue.map(
+                    "add",
+                    [
+                        {"a": 1, "b": 2},
+                        {"a": 3, "b": 4},
+                    ]
+                ) == [3, 7]
+            except JobError:
+                print("any of the jobs failed")
+
+        job_or_func: Same as Queue.enqueue
+        iter_kwargs: Enqueue a job for each item in this sequence. Each item is the same
+            as kwargs for Queue.enqueue.
+        timeouts: Total seconds to wait for all jobs to complete. If None or 0, wait forever.
+        kwargs: Default kwargs for all jobs. These will be overridden by those in iter_kwargs.
+        """
+        if timeout is None:
+            timeout = float("inf")
+        expires = time.time() + timeout
+
+        jobs = [
+            await self.enqueue(job_or_func, **{**kwargs, **kw}) for kw in iter_kwargs
+        ]
+
+        results = []
+        for job in jobs:
+            wait_timeout = expires - time.time()
+            if wait_timeout <= 0:
+                raise asyncio.TimeoutError()
+
+            results.append(await job.wait_for_result(wait_timeout))
+
+        return results
+
+    @asynccontextmanager
+    async def cancel_scope(self):
```

Comment:
i wonder if the name would be more clear as a batch or something 

async with queue.batch():

Reply:
Yeah, I struggled with the name here.

`batch` is nice and simple. I'll switch to that

Reply:
Hm, more thoughts here.

`Batch` feels like an _instance_ of something to me. For example, it seems reasonable to have to simultaneous batches.

In the way it's implemented here, you can only stack these scopes.

Reply:
batch can be thought of as a verb, like batch things